src/backend/distributed/utils/statistics_collection.c

```diff
@@ -0,0 +1,157 @@
+
+static uint64_t NextPow2(uint64_t n);
+static uint64_t ClusterSize(List *distributedTableList);
+static bool SendHttpPostRequest(const char *url, const char *postFields, long timeout_seconds);
+
+/*
+ * CollectBasicUsageStatistics sends basic usage statistics to Citus servers.
+ * This includes Citus version, table count rounded to next power of 2, cluster
+ * size rounded to next power of 2, worker node count, and uname data.
+ */
+void
+CollectBasicUsageStatistics(void)
+{
+	List *distributedTables = NIL;
+	uint64_t roundedDistTableCount = 0;
+	uint64_t roundedClusterSize = 0;
+	uint32_t workerNodeCount = 0;
+	const long timeout_seconds = 5;
+	struct utsname unameData;
+	StringInfo fields = makeStringInfo();
+
+	StartTransactionCommand();
+	distributedTables = DistributedTableList();
+	roundedDistTableCount = NextPow2(list_length(distributedTables));
+	roundedClusterSize = NextPow2(ClusterSize(distributedTables));
+	workerNodeCount = ActivePrimaryNodeCount();
+	CommitTransactionCommand();
+
+	uname(&unameData);
+
+	appendStringInfo(fields, "citus_version=%s", CITUS_VERSION);
+	appendStringInfo(fields, "&table_count=" UINT64_FORMAT, roundedDistTableCount);
+	appendStringInfo(fields, "&cluster_size=" UINT64_FORMAT, roundedClusterSize);
+	appendStringInfo(fields, "&worker_node_count=%u", workerNodeCount);
+	appendStringInfo(fields, "&os_name=%s&os_release=%s&hwid=%s",
+					 unameData.sysname, unameData.release, unameData.machine);
+
+	SendHttpPostRequest("http://localhost:5000/collect_stats", fields->data, timeout_seconds);
+}
+
+
+/*
+ * ClusterSize returns total size of data store in the cluster consisting of
+ * given distributed tables. We ignore tables which we cannot get their size.
+ */
+static uint64_t
+ClusterSize(List *distributedTableList)
+{
+	uint64_t clusterSize = 0;
+	ListCell *distTableCacheEntryCell = NULL;
+
+	foreach(distTableCacheEntryCell, distributedTableList)
+	{
+		DistTableCacheEntry *distTableCacheEntry = lfirst(distTableCacheEntryCell);
+		Oid relationId = distTableCacheEntry->relationId;
+
+		PG_TRY();
+		{
+			Datum distTableSizeDatum = DirectFunctionCall1(citus_table_size,
+														   ObjectIdGetDatum(relationId));
+			clusterSize += DatumGetInt64(distTableSizeDatum);
+		}
+		PG_CATCH();
+		{
+			FlushErrorState();
+		}
+		PG_END_TRY();
+	}
+
+	return clusterSize;
+}
+
+
+/*
+ * NextPow2 returns smallest power of 2 less than or equal to n. If n is greater
+ * than 2^63, it returns 2^63.
+ */
+static uint64_t
+NextPow2(uint64_t n)
+{
+	uint64_t result = 1;
+
+	/* if there is no 64-bit power of 2 greater than n, return 2^63 */
+	if (n > (1ull << 63))
+	{
+		return (1ull << 63);
+	}
+
+	while (result < n)
+	{
+		result *= 2;
+	}
+
+	return result;
+}
+
+
+/*
+ * SendHttpPostRequest sends a HTTP/HTTPS POST request to the given URL with the
+ * given POST fields.
+ */
+static bool
+SendHttpPostRequest(const char *url, const char *postFields, long timeout_seconds)
```

Comment:
What's the point of the return value here?

If the first request fails, that's all we try for an entire day, right? We always update the most recent submission timestamp, even if no data was sent?

Reply:
OK, seems like an OK thing to start with (though we run the risk of self-DDOSing every minute if we get a lot of traffic and are down).